src/grain/ledger.js

```diff
@@ -0,0 +1,220 @@
+ * A grainholder may transfer grain to themself (which is a no-op).
+ */
+export type Transfer = {|
+  +type: "TRANSFER",
+  +version: string,
+  +sender: NodeAddressT,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+/**
+ * Models a participant recieving newly minted grain.
+ */
+export type Distribution = {|
+  +type: "DISTRIBUTION",
+  +version: string,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type LedgerEvent = Distribution | Transfer;
+
+/**
+ * A ledger tracks balances, earnings, and the event history of participants.
+ *
+ * We may refactor the `events` to be an async generator in the future,
+ * when we stop storing all the events in memory.
+ */
+export interface Ledger {
+  /**
+   * Stores the current grain balance of each address.
+   */
+  balances(): Map<NodeAddressT, Grain>;
+  /**
+   * Stores the lifetime earnings of each address.
+   *
+   * Necessary for computing which participants have been underpaid.
+   */
+  earnings(): Map<NodeAddressT, Grain>;
+  /**
+   * Retrieve the events.
+   *
+   * Can be filtered and sorted based on the
+   * EventsOptions, see that type for documentation.
+   */
+  events(?EventsOptions): $ReadOnlyArray<LedgerEvent>;
+}
+
+export class InMemoryLedger implements Ledger {
+  _balances: Map<NodeAddressT, Grain>;
+  _earnings: Map<NodeAddressT, Grain>;
+  _events: $ReadOnlyArray<LedgerEvent>;
+
+  /**
+   * Construct a Ledger from an array of in-memory events.
+   *
+   * The events must be in timestamp sorted order.
+   */
+  constructor(events: $ReadOnlyArray<LedgerEvent>) {
+    this._balances = new Map();
+    this._earnings = new Map();
+    this._events = events;
+    let lastTimestamp = -Infinity;
+    for (const e of events) {
+      if (e.timestampMs < lastTimestamp) {
+        throw new Error(
+          `event timestamps out of order: encountered ${e.timestampMs} after ${lastTimestamp}`
+        );
+      }
+      switch (e.type) {
+        case "DISTRIBUTION":
+          this._processDistribution((e: Distribution));
+          break;
+        case "TRANSFER":
+          this._processTransfer((e: Transfer));
+          break;
+        default:
+          throw new Error(`Unsupported event type: ${(e.type: empty)}`);
+      }
+    }
+  }
+
+  events(options: ?$Shape<EventsOptions>): $ReadOnlyArray<LedgerEvent> {
+    const fullOptions = {
+      ...DEFAULT_EVENTS_OPTIONS,
+      ...NullUtil.orElse(options, {}),
+    };
+    const {address, eventOrder, eventSort} = fullOptions;
+    let filter = (_) => true;
+    if (address) {
+      filter = (x: LedgerEvent) => {
+        switch (x.type) {
+          case "DISTRIBUTION":
+            return x.recipient === address;
+          case "TRANSFER":
+            return x.recipient === address || x.sender === address;
+          default:
+            throw new Error((x.type: empty));
+        }
+      };
+    }
+    const events = this._events.filter(filter);
```

Comment:
Nit: I think it's worth commenting that this variable is expected to be a shallow copy.
In the case we're not calling `sortBy()`, but do call `.reverse()` it would be updated in-place.

Reply:
I've removed the query methods on the ledger, to save them for when they're actually motivated by a use case.